### PR TITLE
ceph-volume: add dmcrypt support in raw mode

### PR DESCRIPTION
**Pull request by guits**
This commit adds the dmcrypt support in `ceph-volume raw` mode.

Fixes: https://tracker.ceph.com/issues/45959

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

### CODE REVIEW

**Comment by guits** [comment]
@leseb

**Comment by leseb** [comment]
Please attach the tracker bug for the backport: https://tracker.ceph.com/issues/45959

File: src/ceph-volume/ceph_volume/devices/raw/activate.py [hunk 1]
```diff
@@ -4,14 +4,15 @@
 import os
```
Comment (guits) [inline src/ceph-volume/ceph_volume/devices/raw/activate.py:4]: @leseb I think all modifications in this file should be removed given we don't need the activation part, right?

Comment (leseb) [reply src/ceph-volume/ceph_volume/devices/raw/activate.py:4]: Yup

File: src/ceph-volume/ceph_volume/devices/raw/prepare.py [hunk 2]
```diff
@@ -1,15 +1,39 @@
...
+        key,
+        device
+    )
+    encryption_utils.luks_open(
+        key,
+        device,
+        mapping
+    )
+
+    return '/dev/mapper/%s' % mapping
```
Comment (jan--f) [inline src/ceph-volume/ceph_volume/devices/raw/prepare.py:36]: Nit, but I'd prefer to use `'/dev/mapper/{}'.format(mapping)`.

File: src/ceph-volume/ceph_volume/devices/raw/prepare.py [hunk 3]
```diff
@@ -94,6 +134,8 @@ def prepare(self):
         # reuse a given ID if it exists, otherwise create a new ID
         self.osd_id = prepare_utils.create_id(
             osd_fsid, json.dumps(secrets))
+        data = self.args.data
```
Comment (jan--f) [inline src/ceph-volume/ceph_volume/devices/raw/prepare.py:137]:
```suggestion
```

Seems like data is never used.

File: src/ceph-volume/ceph_volume/devices/raw/prepare.py [hunk 4]
```diff
@@ -132,5 +172,8 @@ def main(self):
         if not self.args.bluestore:
             terminal.error('must specify --bluestore (currently the only supported backend)')
             raise SystemExit(1)
+        if self.args.dmcrypt and not os.getenv('CEPH_VOLUME_DMCRYPT_SECRET'):
+            terminal.error('Could not read value from environment variable CEPH_VOLUME_DMCRYPT_SECRET, you must set this variable.')
```
Comment (jan--f) [inline src/ceph-volume/ceph_volume/devices/raw/prepare.py:176]:
```suggestion
            terminal.error('Could not read value from environment variable CEPH_VOLUME_DMCRYPT_SECRET, you must set this variable when passing --dmcrypt.')
```

Maybe lets print a bit more explicit error message?

File: src/ceph-volume/ceph_volume/devices/raw/prepare.py [hunk 5]
```diff
@@ -79,6 +111,15 @@ def get_cluster_fsid(self):
...
     def prepare(self):
         secrets = {'cephx_secret': prepare_utils.create_key()}
+        encrypted = 1 if self.args.dmcrypt else 0
+        cephx_lockbox_secret = '' if not encrypted else prepare_utils.create_key()
+
+        if encrypted:
+            secrets['dmcrypt_key'] = os.getenv('CEPH_VOLUME_DMCRYPT_SECRET')
+            secrets['cephx_lockbox_secret'] = cephx_lockbox_secret # dummy value to make `ceph osd new` not complaining
+
+        cluster_fsid = self.get_cluster_fsid()
```
Comment (guits) [inline src/ceph-volume/ceph_volume/devices/raw/prepare.py:121]: not used

File: src/ceph-volume/ceph_volume/devices/raw/prepare.py [hunk 6]
```diff
@@ -64,21 +94,20 @@ def safe_prepare(self, args=None):
...
             raise
-        terminal.success("ceph-volume raw prepare successful for: %s" % self.args.data)
-
-    def get_cluster_fsid(self):
-        """
-        Allows using --cluster-fsid as an argument, but can fallback to reading
-        from ceph.conf if that is unset (the default behavior).
-        """
-        if self.args.cluster_fsid:
-            return self.args.cluster_fsid
```
Comment (jan--f) [inline src/ceph-volume/ceph_volume/devices/raw/prepare.py:75]: Is this code meant to be removed? The option is still offered by the cli (see https://github.com/ceph/ceph/blob/79104332d98d5857fc74285320102fc232e6e0ed/src/ceph-volume/ceph_volume/devices/raw/common.py#L29)?
Seems like it was never used, so maybe also drop the cli arg?

**Comment by jan--f** [comment]
jenkins test ceph-volume tox

**Comment by jan--f** [comment]
jenkins test ceph-volume tox

**Comment by guits** [comment]
jenkins test ceph-volume tox

**Comment by JINHAOHU** [comment]
Why we don't need to decrypt the device during the activate process? I see that during the activate process of the lvm mode,  it will call [LUKS open](https://github.dev/ceph/ceph/blob/5ab5ff46d6d799c46d8a1c01cc08068a22e4c9b5/src/ceph-volume/ceph_volume/devices/lvm/activate.py#L176-L177) to decrypt the device.

**Pull request merged**